CMakeLists.txt

```diff
@@ -485,6 +485,33 @@ else()
     endif()
   endif()
 
+  option(NO_AES "Explicitly disable AES support" ${NO_AES})
+  if(NO_AES)
+    message(STATUS "AES support explicitly disabled")
+    add_definitions("-DNO_AES")
+  elseif(NOT ARM AND NOT PPC64LE AND NOT PPC64 AND NOT PPC AND NOT S390X)
+    message(STATUS "AES support enabled")
+    add_compile_options("-maes")
+  elseif(PPC64LE OR PPC64 OR PPC)
+    message(STATUS "AES support not available on POWER")
+  elseif(S390X)
+    message(STATUS "AES support not available on s390x")
+  elseif(ARM6)
+    message(STATUS "AES support not available on ARMv6")
+  elseif(ARM7)
+    message(STATUS "AES support not available on ARMv7")
+  elseif(ARM8)
+    CHECK_CXX_ACCEPTS_FLAG("-march=${ARCH}+crypto" ARCH_PLUS_CRYPTO)
+    if(ARCH_PLUS_CRYPTO)
+      message(STATUS "Crypto extensions enabled for ARMv8")
+      set(ARCH_FLAG "-march=${ARCH}+crypto")
```

Comment:
This adding +crypto to ARCH was (and is) wrong: the only time `ARCH` is set to something other than `native` is if someone explicitly specified it, and if they explicitly specified `-DARCH=armv8-a` then it is very wrong to actually compile with `-march=armv8-a+crypto` which results in a binary that does *not* work on the ARCH value they specified.  Adding +crypto would be the x86 equivalent of me specifying `ARCH=haswell` and the build system saying "oh you said you want to target haswell, but skylake works on the host system, so I'm going to change it on you."

If you want crypto extensions you should be building with `-DARCH=armv8-a+crypto` instead of `-DARCH=armv8-a` (and indeed, this is what you get with the default `ARCH=native` *if* you are on a crypto extension supporting system).

But for a generic ARM64 build you definitely do *not* want crypto extensions because there are lots of common ARM devices (like Raspberry Pis, even the latest Pi 4) where the resulting binary will not work properly because crypto extensions are not a required ARM64 CPU feature.  (The only reason we didn't hit this before is that gcc doesn't seem to use the +crypto for general optimizations, and android devices never had a need to call into the turtle hashing code).

> enables `vaeseq_u8` required for the PoW algorithms on armv8.

Those are only required *if* compiling with `+crypto`.  Those only look to be required if compiling turtle hash with +crypto, as far as I can see; when not compiling with crypto there's an alternate code path.

Reply:
Also I opened PR #1099 to just drop this block of cmake code entirely, as upon closer inspection it serves no useful purpose at all that isn't already done properly with the ARCH=native default cmake option.

Reply:
Removing this means we should rework the preprocessor macros to properly stub/wrapper the extension functions in the PoW files (seems Heavy Hash, even RandomX were written assuming crypto extensions). They implicitly rely on +crypto extension functionality for linkage but, the reason it all works even if the CPU does not have AES extension is because they switch to the software implementation at run-time if the CPU fails the AES feature flag.

Turtle actually seems ok considering they use the `__ARM_FEATURE_CRYPTO` macro. We can remove and I can add temporarily the Android build script because we're cross-compiling, but to fix the underlying problem needs some more work.

I agree with everything you said, doesn't make sense to override the user's option.

Reply:
Yeah, RandomX has a much better design here than cn_turtle did.

(Lokinet does something very similar -- we always try to build optimized AVX2 code (when not doing everything with AVX2) and replace the function pointers at runtime *if* the CPU indicates AVX2 support at runtime).

> They implicitly rely on +crypto extension functionality for linkage

I haven't had any errors with heavy or turtle linking on my Pi (I restarted a build earlier with #1099 applied and `-DARCH=archv8-a+crc`).  I definitely can't build on stretch or xenial though -- both give compilation errors for some intrinsic arm functions that aren't supported on gcc 5 (xenial) or 6 (stretch).

Reply:
-DARCH=armv8-a` should trigger the compiler to complain with `vaseq_u8`.

Reply:
God dammit:
```
#if defined(__aarch64__)
#ifndef __clang__
#pragma GCC target ("+crypto")
#endif
```

Reply:
Oh, the cn_heavy one actually looks okay: it does a hardware check for AES instruction support at runtime.